### PR TITLE
Add com.chenyifaer.FaForever

### PR DESCRIPTION
**Pull request by kjxbyz**
Name: FaForever

A third-party music player that replaces Synology DS Audio client.

### Please confirm your submission meets all the criteria

- [X] Please describe your application briefly.
- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I have [built][build] and tested the submission locally.
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

### CODE REVIEW

File: com.chenyifaer.FaForever.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        <p>A third-party music player that replaces Synology DS Audio client.</p>
+        <p>Features</p>
+        <ul>
+            <li>Open Source!</li>
+        </ul>
+    </description>
+    <launchable type="desktop-id">com.chenyifaer.FaForever.desktop</launchable>
+    <screenshots>
+        <screenshot type="default">
+            <image>https://raw.githubusercontent.com/cyf/faforever/main/screenshots/desktop.png</image>
```
Comment (hfiguiere) [inline com.chenyifaer.FaForever.appdata.xml:31]: you should never reference a branch.

(same with the other screenshots)

File: com.chenyifaer.FaForever.yml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    - install -Dm755 usr/bin/fa-forever /app/bin/faforever
+    - mkdir -p /app/share/icons/hicolor/scalable/apps
+    - install -Dm644 com.chenyifaer.FaForever.png /app/share/icons/hicolor/scalable/apps/com.chenyifaer.FaForever.png
+    - install -Dm644 com.chenyifaer.FaForever.desktop /app/share/applications/com.chenyifaer.FaForever.desktop
+    - install -Dm644 com.chenyifaer.FaForever.appdata.xml /app/share/appdata/com.chenyifaer.FaForever.appdata.xml
+    sources:
+      - type: file
+        only-arches:
+        - x86_64
+        url: https://github.com/cyf/faforever/releases/download/v0.0.0+22/fa-forever_0.0.0+22_amd64.deb
```
Comment (hfiguiere) [inline com.chenyifaer.FaForever.yml:70]: it's written "open source" in the appstream file. so why not building from source and offering aarch64 support at the same time.

File: com.chenyifaer.FaForever.png
Comment (hfiguiere) [inline com.chenyifaer.FaForever.png:1]: icon file should be upstream.

File: com.chenyifaer.FaForever.desktop
Comment (hfiguiere) [inline com.chenyifaer.FaForever.desktop:1]: desktop file should be upstream.

File: com.chenyifaer.FaForever.appdata.xml
Comment (hfiguiere) [inline com.chenyifaer.FaForever.appdata.xml:1]: appstream file should be upstream.

File: flathub.json [hunk 3]
```diff
@@ -0,0 +1,5 @@
+{
+    "only-arches": [
+        "x86_64"
```
Comment (hfiguiere) [inline flathub.json:3]: disappointing.

File: com.chenyifaer.FaForever.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  - name: faforever
+    buildsystem: simple
+    only-arches:
+    - x86_64
+    build-commands:
+    - ar -x *.deb
+    - tar -xf data.tar.gz
+    - install -Dm755 usr/bin/fa-forever /app/bin/faforever
+    - mkdir -p /app/share/icons/hicolor/scalable/apps
+    - install -Dm644 com.chenyifaer.FaForever.png /app/share/icons/hicolor/scalable/apps/com.chenyifaer.FaForever.png
```
Comment (hfiguiere) [inline com.chenyifaer.FaForever.yml:63]: PNG files are not scalable.

File: com.chenyifaer.FaForever.yml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  - name: faforever
+    buildsystem: simple
+    only-arches:
+    - x86_64
+    build-commands:
+    - ar -x *.deb
+    - tar -xf data.tar.gz
+    - install -Dm755 usr/bin/fa-forever /app/bin/faforever
+    - mkdir -p /app/share/icons/hicolor/scalable/apps
```
Comment (hfiguiere) [inline com.chenyifaer.FaForever.yml:62]:
```suggestion
```

File: com.chenyifaer.FaForever.yml [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+sdk: org.gnome.Sdk
+command: faforever
+
+finish-args:
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --share=network
+  - --device=dri
+  - --share=ipc
+  - --filesystem=xdg-download
```
Comment (hfiguiere) [inline com.chenyifaer.FaForever.yml:13]: you say it's a music player. I don't see anything for audio playback.

File: com.chenyifaer.FaForever.yml (hunk 2 above)
Comment (kjxbyz) [reply com.chenyifaer.FaForever.yml:70]: flatpak/flatpak-builder-tools#383 blocked me.

File: flathub.json (hunk 3 above)
Comment (kjxbyz) [reply flathub.json:3]: flatpak/flatpak-builder-tools#383 blocked me.

File: com.chenyifaer.FaForever.desktop
Comment (kjxbyz) [reply com.chenyifaer.FaForever.desktop:1]: What does "xxx file should be upstream" mean?

File: com.chenyifaer.FaForever.appdata.xml (hunk 1 above)
Comment (kjxbyz) [reply com.chenyifaer.FaForever.appdata.xml:31]: Can tags be used instead?

File: com.chenyifaer.FaForever.appdata.xml [hunk 7]
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+    Generator for metainfo & .desktop files:
+    https://www.freedesktop.org/software/appstream/metainfocreator/#/
+-->
+<component type="desktop-application">
+    <id>com.chenyifaer.FaForever</id>
+    <name>FaForever</name>
+    <summary>A third-party music player that replaces Synology DS Audio client.</summary>
```
Comment (bbhtt) [inline com.chenyifaer.FaForever.appdata.xml:9]:
```suggestion
    <summary>A third-party music player that replaces Synology DS Audio client</summary>
```

File: com.chenyifaer.FaForever.appdata.xml (hunk 1 above)
Comment (bbhtt) [reply com.chenyifaer.FaForever.appdata.xml:31]: Yea

File: com.chenyifaer.FaForever.desktop
Comment (bbhtt) [reply com.chenyifaer.FaForever.desktop:1]: > What does "xxx file should be upstream" mean?

Here https://github.com/cyf/faforever/

File: com.chenyifaer.FaForever.yml [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+          url-template: https://webkitgtk.org/releases/webkitgtk-$version.tar.xz
+    buildsystem: cmake-ninja
+    config-opts:
+      # taken from https://gitlab.gnome.org/GNOME/gnome-build-meta/-/blob/5a241420c2ee85cc5080dd7cf7ba3049fd80fcf1/elements/world/webkit2gtk-4.0.bst
+      - -DPORT=GTK
+      - -DENABLE_DOCUMENTATION=OFF
+      - -DENABLE_WEBDRIVER=OFF
+      - -DUSE_GTK4=OFF
+      - -DUSE_LIBBACKTRACE=OFF
+      - -DUSE_SOUP2=ON
```
Comment (bbhtt) [inline com.chenyifaer.FaForever.yml:34]: Where is libsoup2?

File: com.chenyifaer.FaForever.desktop
Comment (bbhtt) [reply com.chenyifaer.FaForever.desktop:1]: > What does "xxx file should be upstream" mean?

Here https://github.com/cyf/faforever/

File: com.chenyifaer.FaForever.appdata.xml (hunk 1 above)
Comment (bbhtt) [reply com.chenyifaer.FaForever.appdata.xml:31]: > Can tags be used instead?

Yes

File: com.chenyifaer.FaForever.yml [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+finish-args:
+  - --socket=fallback-x11
+  - --socket=wayland
+  - --share=network
+  - --device=dri
+  - --share=ipc
+  - --filesystem=xdg-download
+
+modules:
+  - name: webkit2gtk-4.0
```
Comment (bbhtt) [inline com.chenyifaer.FaForever.yml:16]: This should really be ported to use newer webkit2gtk. What needs 4.0?

File: com.chenyifaer.FaForever.yml (hunk 8 above)
Comment (kjxbyz) [reply com.chenyifaer.FaForever.yml:34]: libsoup2 is imported on line 36

File: com.chenyifaer.FaForever.yml (hunk 9 above)
Comment (kjxbyz) [reply com.chenyifaer.FaForever.yml:16]: Reference this

https://github.com/flathub/in.cinny.Cinny/blob/6ff4f0273e4a5edcbb00cd7faca09e89071c0539/in.cinny.Cinny.yml#L32

File: com.chenyifaer.FaForever.yml (hunk 8 above)
Comment (bbhtt) [reply com.chenyifaer.FaForever.yml:34]: Indeed, missed it.

File: com.chenyifaer.FaForever.yml (hunk 9 above)
Comment (bbhtt) [reply com.chenyifaer.FaForever.yml:16]: Is this tauri too?

File: com.chenyifaer.FaForever.yml (hunk 6 above)
Comment (kjxbyz) [reply com.chenyifaer.FaForever.yml:13]: Is there some value missing in `finish-args`?

File: com.chenyifaer.FaForever.yml (hunk 9 above)
Comment (kjxbyz) [reply com.chenyifaer.FaForever.yml:16]: Yes

File: flathub.json (hunk 3 above)
Comment (hfiguiere) [reply flathub.json:3]: maybe you can help and contribute.

File: com.chenyifaer.FaForever.yml (hunk 6 above)
Comment (hfiguiere) [reply com.chenyifaer.FaForever.yml:13]: you tell me. But if this is supposed to play sound, then you likely never tested the package.

File: flathub.json (hunk 3 above)
Comment (kjxbyz) [reply flathub.json:3]: I now spend all my spare time developing the build and distribution package process, and I don't have any spare time left😭, sometimes I even forget to eat😭😭.

File: com.chenyifaer.FaForever.desktop
Comment (kjxbyz) [reply com.chenyifaer.FaForever.desktop:1]: Why must these files be in the upstream? I see other packages are the same as mine. Can you tell me the reason?

Comment (hfiguiere) [reply com.chenyifaer.FaForever.desktop:1]: two wrongs don't make a right.

They are not specific to flatpak, they should be managed with the upstream source. The lack of .desktop for a desktop app is a BUG. (the spec is more than 20 year old)

**Comment by kjxbyz** [comment]
@bbhtt Updated.

File: com.chenyifaer.FaForever.yml [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      - install -Dm644 usr/share/icons/hicolor/32x32/apps/fa-forever.png /app/share/icons/hicolor/32x32/apps/com.chenyifaer.FaForever.png
+      - install -Dm644 usr/share/icons/hicolor/256x256@2/apps/fa-forever.png /app/share/icons/hicolor/256x256@2/apps/com.chenyifaer.FaForever.png
+      - install -Dm644 com.chenyifaer.FaForever.appdata.xml /app/share/appdata/com.chenyifaer.FaForever.appdata.xml
+    sources:
+      - type: file
+        only-arches:
+          - x86_64
+        url: https://github.com/cyf/faforever/releases/download/v0.0.1+24/fa-forever_0.0.1+24_amd64.deb
+        sha256: 73a741b70e64c67090ac5fca860b306526cbca2dd1c913c9a7ad5efec35e7b12
+        x-checker-data:
```
Comment (bbhtt) [inline com.chenyifaer.FaForever.yml:85]:
```suggestion
      - type: file
        only-arches:
          - x86_64
        url: https://github.com/cyf/faforever/releases/download/v0.0.1+24/fa-forever_0.0.1+24_amd64.deb
        sha256: 73a741b70e64c67090ac5fca860b306526cbca2dd1c913c9a7ad5efec35e7b12
        dest-filename: fa-forever.deb
        x-checker-data:
```

File: com.chenyifaer.FaForever.yml [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  # for some controller mods (see https://github.com/isXander/Controlify/issues/31)
+  - shared-modules/libusb/libusb.json
+
+  - name: faforever
+    buildsystem: simple
+    only-arches:
+      - x86_64
+    build-commands:
+      - ar -x *.deb
+      - tar -xf data.tar.gz
```
Comment (bbhtt) [inline com.chenyifaer.FaForever.yml:72]:
```suggestion
      - bsdtar -Oxf fa-forever.deb data.tar.xz | bsdtar -xf -
```

Comment (kjxbyz) [reply com.chenyifaer.FaForever.yml:72]: The command should be `bsdtar -Oxf fa-forever.deb data.tar.xz | bsdtar -xf -`. In addition, does the runtime include the `bsdtar` command?

Comment (bbhtt) [reply com.chenyifaer.FaForever.yml:72]: Yes it does snd that's what I wrote

Comment (bbhtt) [reply com.chenyifaer.FaForever.yml:72]: Ah the it is a gzip here fixed

**Comment by kjxbyz** [comment]
@bbhtt Fixed

**Comment by bbhtt** [comment]
bot, build com.chenyifaer.FaForever

File: com.chenyifaer.FaForever.yml [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+      - type: file
+        url: https://raw.githubusercontent.com/cyf/faforever/v0.0.1+24/linux/com.chenyifaer.FaForever.desktop
+        dest-filename: com.chenyifaer.FaForever.desktop
+        sha256: eeb8e140a8b78911dd22c3a09dde11cfd0ff3d14c1bc0ab2732ca05c0acb78fb
+
+      - type: file
+        url: https://raw.githubusercontent.com/cyf/faforever/v0.0.1+24/linux/com.chenyifaer.FaForever.appdata.xml
+        dest-filename: com.chenyifaer.FaForever.appdata.xml
+        sha256: c6b985d6acb05a13164a7a727b5725da7b72f049dc8ef2fdb5325c8d84d5f418
```
Comment (bbhtt) [inline com.chenyifaer.FaForever.yml:99]: Screenshot links are 404

Comment (kjxbyz) [reply com.chenyifaer.FaForever.yml:99]: Yes, I forgot to upload the screenshot, I am working on it

**Comment by bbhtt** [comment]
bot, build com.chenyifaer.FaForever

**Comment by kjxbyz** [comment]
Sorry, forgot to update sha256, which caused the build to fail! 

**Now it has been fixed!!!** 

By the way, how can I update the sha256 of a file while updating the file version?

**Comment by hfiguiere** [comment]
by using `sha256sum` on the file.

**Comment by bbhtt** [comment]
x-checker will update version and hash both when there is an update. Otherwise you have to manually do it.

bot, build com.chenyifaer.FaForever

**Comment by bbhtt** [comment]
image


The play button opens a link that is 404 and the create window seems a leftover? It opens some dev docs thing for Tauri. I do not see any indication of how this is a music player?

**Comment by kjxbyz** [comment]
Music playback will be added soon, and unnecessary features will be removed.

**Pull request closed without merging**